src/pip/_internal/operations/prepare.py

```diff
@@ -174,7 +170,7 @@ def get_file_url(
 
 def unpack_url(
     link: Link,
-    location: str,
+    location: Optional[str],
```

Comment:
I don't think this should be None in any cases; the assertion should probably be done at the caller site.

Reply:
Hmm, according to test `tests/functional/test_build_env.py::test_build_env_allow_only_one_install` this value can be None when:

```
link.url not in downloaded
link.is_vcs=False
link.is_existing_dir()=False
link.is_file=True
link.is_wheel=True
```

During this test, `req.source_dir` is None.

```
Traceback (most recent call last):
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/cli/base_command.py", line 179, in exc_logging_wrapper
    status = run_func(*args)
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/cli/req_command.py", line 203, in wrapper
    return func(self, options, args)
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/commands/install.py", line 340, in run
    requirement_set = resolver.resolve(
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/resolution/resolvelib/resolver.py", line 94, in resolve
    result = self._result = resolver.resolve(
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_vendor/resolvelib/resolvers.py", line 472, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_vendor/resolvelib/resolvers.py", line 341, in resolve
    self._add_to_criteria(self.state.criteria, r, parent=None)
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_vendor/resolvelib/resolvers.py", line 172, in _add_to_criteria
    if not criterion.candidates:
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_vendor/resolvelib/structs.py", line 151, in __bool__
    return bool(self._sequence)
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/resolution/resolvelib/found_candidates.py", line 155, in __bool__
    return any(self)
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/resolution/resolvelib/found_candidates.py", line 143, in <genexpr>
    return (c for c in iterator if id(c) not in self._incompatible_ids)
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/resolution/resolvelib/found_candidates.py", line 47, in _iter_built
    candidate = func()
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/resolution/resolvelib/factory.py", line 201, in _make_candidate_from_link
    self._link_candidate_cache[link] = LinkCandidate(
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/resolution/resolvelib/candidates.py", line 281, in __init__
    super().__init__(
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/resolution/resolvelib/candidates.py", line 156, in __init__
    self.dist = self._prepare()
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/resolution/resolvelib/candidates.py", line 225, in _prepare
    dist = self._prepare_distribution()
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/resolution/resolvelib/candidates.py", line 292, in _prepare_distribution
    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/operations/prepare.py", line 492, in prepare_linked_requirement
    return self._prepare_linked_requirement(req, parallel_builds)
  File "/tmp/pip-standalone-pip-z_myip83/__env_pip__.zip/pip/_internal/operations/prepare.py", line 539, in _prepare_linked_requirement
    assert req.source_dir is not None
AssertionError
```

Do you think this is a non-representative test or that `location` can in fact be `None`?